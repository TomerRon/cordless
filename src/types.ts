import { ApplicationCommandOptionAllowedChannelTypes } from '@discordjs/builders'
import { APIApplicationCommandOptionChoice } from 'discord-api-types/v10'
import Discord, {
  ApplicationCommandOptionType,
  ClientEvents,
  ClientOptions,
  CommandInteraction,
} from 'discord.js'

/** Initialization options for your cordless bot */
export type InitOptions<C extends CustomContext = {}> = {
  /** The commands used by your bot. */
  commands?: BotCommand<C>[]
  /** The functions used by your bot */
  functions: BotFunction<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
  /**
   * Your bot token.
   *
   * @see https://discordjs.guide/preparations/setting-up-a-bot-application.html#your-bot-s-token
   */
  token: string
  /** A custom context object which will extend the context passed to your commands and functions */
  context?: C
  /**
   * Override the default Gateway Intents of the discord.js client.
   *
   * By default, the discord.js client will initialize with the [GUILDS, GUILD_MESSAGES] intents.
   * These default intents should be sufficient in most cases.
   *
   * @see https://discord.com/developers/docs/topics/gateway#gateway-intents
   * @see https://discordjs.guide/popular-topics/intents.html
   */
  intents?: ClientOptions['intents']
  /**
   * Generate a help function for the bot, which will be triggered by the value.
   *
   * For example, given a value of "!help", the following commands will be available:
   *
   * - !help (shows a list of available functions)
   * - !help <function-name> (shows a description and usage instructions for the given function)
   */
  helpCommand?: string
}

export type BotCommand<C extends CustomContext = {}> =
  | BotCommandWithHandler<C>
  | BotCommandWithSubcommands<C>

export interface BotCommandWithHandler<C extends CustomContext = {}>
  extends BotCommandBase {
  handler: (args: BotCommandHandlerArgs<C>) => void | Promise<void>
  options?: BotCommandOption[]
  subcommands?: never
}

export interface BotCommandWithSubcommands<C extends CustomContext = {}>
  extends BotCommandBase {
  subcommands: BotCommandWithHandler<C>[]
  handler?: never
  options?: never
}

type BotCommandBase = {
  name: string
  description?: string
}

export type BotCommandHandlerArgs<C extends CustomContext = {}> = {
  interaction: CommandInteraction
  context: Context<C>
}

export type BotCommandOption =
  | BotCommandStringOption
  | BotCommandIntegerOption
  | BotCommandBooleanOption
  | BotCommandUserOption
  | BotCommandChannelOption
  | BotCommandRoleOption
  | BotCommandMentionableOption
  | BotCommandNumberOption
  | BotCommandAttachmentOption

export interface BotCommandStringOption extends BotCommandOptionBase {
  type: 'STRING'
  choices?: APIApplicationCommandOptionChoice<string>[]
}

export interface BotCommandIntegerOption extends BotCommandOptionBase {
  type: 'INTEGER'
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandBooleanOption extends BotCommandOptionBase {
  type: 'BOOLEAN'
}

export interface BotCommandUserOption extends BotCommandOptionBase {
  type: 'USER'
}

export interface BotCommandChannelOption extends BotCommandOptionBase {
  type: 'CHANNEL'
  channelTypes?: ApplicationCommandOptionAllowedChannelTypes[]
}

export interface BotCommandRoleOption extends BotCommandOptionBase {
  type: 'ROLE'
}

export interface BotCommandMentionableOption extends BotCommandOptionBase {
  type: 'MENTIONABLE'
}

export interface BotCommandNumberOption extends BotCommandOptionBase {
  type: 'NUMBER'
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandAttachmentOption extends BotCommandOptionBase {
  type: 'ATTACHMENT'
}

type BotCommandOptionBase = {
  type: ApplicationCommandOptionType
  name: string
  description?: string
  required?: boolean
}

export type BotFunction<
  E extends keyof ClientEvents = 'messageCreate',
  C extends CustomContext = {},
> = {
  /**
   * The name of this function (no spaces allowed).
   * It will be displayed in the help function if there is a helpCommand.
   */
  name?: string
  /**
   * The description of this function and usage instructions.
   * It will be displayed in the help function if there is a helpCommand.
   */
  description?: string
  /**
   * The event this function should subscribe to (default: "messageCreate").
   */
  event?: E
  /** Determines whether or not this function should run */
  condition: (...args: [...ClientEvents[E], Context<C>]) => boolean
  /** Called whenever this function should run */
  callback: (
    ...args: [...ClientEvents[E], Context<C>]
  ) => void | Promise<Discord.Message | void>
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomContext = Record<string, any>

export type Context<C extends CustomContext = {}> = {
  client: Discord.Client<true>
  functions: BotFunction<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
} & C
