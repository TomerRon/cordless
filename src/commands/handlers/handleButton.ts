import { ButtonInteraction } from 'discord.js'
import { BotCommandButtonHandler, Context, CustomContext } from '../../types'

type HandleButtonArgs<C extends CustomContext> = {
  buttonHandlerMap: Record<string, BotCommandButtonHandler<C>>
  interaction: ButtonInteraction
  context: Context<C>
}

const handleButton = async <C extends CustomContext = {}>({
  buttonHandlerMap,
  interaction,
  context,
}: HandleButtonArgs<C>): Promise<void> => {
  const handler = buttonHandlerMap[interaction.customId]

  if (!handler) return

  await handler({ interaction, context })
}

export default handleButton
