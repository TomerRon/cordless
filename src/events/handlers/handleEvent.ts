import { ClientEvents } from 'discord.js'
import { BotEventHandler, Context, CustomContext } from '../../types'
import isSelfEvent from '../utils/isSelfEvent'

const handleEvent = async <
  E extends keyof ClientEvents,
  C extends CustomContext,
>(
  eventArgs: ClientEvents[E],
  handlers: BotEventHandler<E, C>[],
  context: Context<C>,
): Promise<void> => {
  const { client } = context

  if (!client.user) {
    return
  }

  if (isSelfEvent(eventArgs, client.user.id)) {
    return
  }

  const result = handlers.find(({ condition }) =>
    condition(...eventArgs, context),
  )

  await result?.callback(...eventArgs, context)
}

export default handleEvent
