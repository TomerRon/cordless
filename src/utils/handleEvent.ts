import { ClientEvents } from 'discord.js'
import { BotFunction, Context, CustomContext } from '../types'
import isSelfEvent from './isSelfEvent'

const handleEvent = async <
  E extends keyof ClientEvents,
  C extends CustomContext,
>(
  eventArgs: ClientEvents[E],
  functions: BotFunction<E, C>[],
  context: Context<C>,
): Promise<void> => {
  const { client } = context

  if (!client.user) {
    return
  }

  if (isSelfEvent(eventArgs, client.user.id)) {
    return
  }

  const result = functions.find(({ condition }) =>
    condition(...eventArgs, context),
  )

  await result?.callback(...eventArgs, context)
}

export default handleEvent
