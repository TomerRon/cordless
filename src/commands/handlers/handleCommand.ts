import { ChatInputCommandInteraction } from 'discord.js'
import { BotCommand, Context, CustomContext } from '../../types'
import buildComponents from '../builders/buildComponents'
import { isCommandWithSubcommands } from '../utils/guards'

type HandleCommandArgs<C extends CustomContext> = {
  commands: BotCommand<C>[]
  interaction: ChatInputCommandInteraction
  context: Context<C>
}

const handleCommand = async <C extends CustomContext = {}>({
  commands,
  interaction,
  context,
}: HandleCommandArgs<C>): Promise<void> => {
  let command = commands.find(({ name }) => interaction.commandName === name)

  if (!command) {
    return
  }

  if (isCommandWithSubcommands(command)) {
    const subcommandName = interaction.options.getSubcommand()

    const subcommand = command.subcommands.find(
      ({ name }) => subcommandName === name,
    )

    if (!subcommand) {
      return
    }

    command = subcommand
  }

  const components = await buildComponents({ command, interaction, context })

  await command.handler({
    context,
    interaction,
    components,
  })
}

export default handleCommand
