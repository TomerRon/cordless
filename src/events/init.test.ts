import { Client, Message } from 'discord.js'
import { BotFunction, Context } from '../types'
import initEvents, { InitEventsArgs } from './init'
import * as handleEventModule from './handlers/handleEvent'

describe('initEvents', () => {
  // A function without an explicit event will map to a messageCreate handler
  const fnA: BotFunction = {
    condition: jest.fn(),
    callback: jest.fn(),
  }

  const fnB: BotFunction<'channelCreate'> = {
    event: 'channelCreate',
    condition: jest.fn(),
    callback: jest.fn(),
  }

  const fnC: BotFunction<'messageCreate'> = {
    event: 'messageCreate',
    condition: jest.fn(),
    callback: jest.fn(),
  }

  const fnD: BotFunction<'messageDelete'> = {
    event: 'messageDelete',
    condition: jest.fn(),
    callback: jest.fn(),
  }

  const fnE: BotFunction<'channelCreate'> = {
    event: 'channelCreate',
    condition: jest.fn(),
    callback: jest.fn(),
  }

  const mockFunctions = [fnA, fnB, fnC, fnD, fnE]

  const mockClient = {
    on: jest.fn(),
  } as unknown as Client<true>

  const mockContext: Context = { client: mockClient, functions: [] }

  const mockArgs: InitEventsArgs = {
    client: mockClient,
    functions: mockFunctions,
    context: mockContext,
  }

  const mockMsg = {
    content: 'ping',
  } as Message

  const handleEventSpy = jest
    .spyOn(handleEventModule, 'default')
    .mockResolvedValue(undefined)

  it('should handle many functions and map each one to an event handler', async () => {
    initEvents(mockArgs)

    const expectedEvents = {
      messageCreate: [fnA, fnC],
      channelCreate: [fnB, fnE],
      messageDelete: [fnD],
    }

    Object.entries(expectedEvents).forEach(([event, expectedFns], i) => {
      expect(mockClient.on).toHaveBeenNthCalledWith(
        i + 1,
        event,
        expect.any(Function),
      )

      const eventHandler = (mockClient.on as jest.Mock).mock.calls[i][1] as (
        msg: Message,
      ) => void

      eventHandler(mockMsg)

      expect(handleEventSpy).toHaveBeenNthCalledWith(
        i + 1,
        [mockMsg],
        expectedFns,
        mockContext,
      )
    })

    expect(handleEventSpy.mock.calls).toMatchSnapshot()
  })
})
