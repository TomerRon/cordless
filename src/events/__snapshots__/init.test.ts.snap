// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initEvents should handle many functions and map each one to an event handler 1`] = `
Array [
  Array [
    Array [
      Object {
        "content": "ping",
      },
    ],
    Array [
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
      },
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "messageCreate",
      },
    ],
    Object {
      "client": Object {
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "messageCreate",
              [Function],
            ],
            Array [
              "channelCreate",
              [Function],
            ],
            Array [
              "messageDelete",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "functions": Array [],
    },
  ],
  Array [
    Array [
      Object {
        "content": "ping",
      },
    ],
    Array [
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "channelCreate",
      },
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "channelCreate",
      },
    ],
    Object {
      "client": Object {
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "messageCreate",
              [Function],
            ],
            Array [
              "channelCreate",
              [Function],
            ],
            Array [
              "messageDelete",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "functions": Array [],
    },
  ],
  Array [
    Array [
      Object {
        "content": "ping",
      },
    ],
    Array [
      Object {
        "callback": [MockFunction],
        "condition": [MockFunction],
        "event": "messageDelete",
      },
    ],
    Object {
      "client": Object {
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "messageCreate",
              [Function],
            ],
            Array [
              "channelCreate",
              [Function],
            ],
            Array [
              "messageDelete",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "functions": Array [],
    },
  ],
]
`;
